fruitCollection.js

```diff
@@ -0,0 +1,115 @@
+/*
+monkey eating fruit
+tree to tree
+find collection of tree of fruit
+
+list of tree
+each tree: x-coordinate, y-coordinate, num of fruit, vine length
+
+i: [2,4,5,10]
+o: max of fruit 
+
+const trees = [
+    [2, 4, 2, 6],
+    [10, 4, 5, 5],
+    [3, 8, 3, 6],
+    [12, 8, 6, 5],
+    ,[1, 6 2, 6],
+];
+
+*/
+
+const getDistanceBetweenNodes = (nodeOneCoordinates, nodeTwoCoordinates) => {
+    const [nodeOneXcoordinate, nodeOneYCoordinate] = nodeOneCoordinates;
+    const [nodeTwoXcoordinate, nodeTwoYCoordinate] = nodeTwoCoordinates;
+
+    const distance = Math.sqrt(
+        Math.pow(nodeTwoXcoordinate - nodeOneXcoordinate, 2) +
+            Math.pow(nodeTwoYCoordinate - nodeOneYCoordinate, 2),
+    );
+
+    return distance;
+};
+
+const areConnected = (nodeOne, nodeTwo) => {
+    const [nodeOneXcoordinate, nodeOneYCoordinate, nodeOneFruitNum, nodeOneLengthDistance] =
+        nodeOne;
+    const [nodeTwoXcoordinate, nodeTwoYCoordinate, nodeTwoFruitNum, nodeTwoLengthDistance] =
+        nodeTwo;
+
+    const distanceBetweenNodes = getDistanceBetweenNodes(
+        [nodeOneXcoordinate, nodeOneYCoordinate],
+        [nodeTwoXcoordinate, nodeTwoYCoordinate],
+    );
+
+    const smallestDistanceBetweenTwoNodes = Math.min(nodeOneLengthDistance, nodeTwoLengthDistance);
+    return distanceBetweenNodes <= smallestDistanceBetweenTwoNodes;
+};
+
+const buildGraph = (nodes) => {
```

Comment:
P: Suppose I have an array [1,2,3] and I want to generate all pairs of numbers (as you're doing in this function). I could easily get duplicates, like this:
(1,1)
(1,2)
(1,3)
(2,1) -- this is a duplicate!
(2,2)
(2,3)
(3,1) -- this is a duplicate!
(3,2) -- this is a duplicate!
(3,3)

Does your code produce duplicates? Why or why not?

Reply:
Yes, my code could produce duplicates depending on the input array. For the ones that do, it's because it's a bi-directional graph, thus which will give us cycles. To avoid running to the same node, we use a visited set for checks.

Reply:
This is not correct. In fact, your code does NOT produce duplicates.

To be clear, I'm specifically talking about your nested for loops in `buildGraph`. That is, lines 51 and 52 in your current code.

I'd suggest that you copy paste these loops into their own file, and then just add this line inside of the loops:
`console.log(`${nodeOne} ${nodeTwo}`);`

This might give you some further insight into the behavior code. Then, try to figure out why it is the case that your code does NOT produce duplicates in the way I originally described.

Reply:
Ah, I believe I understand. It's because we initialized the first for-loop on the 0-index of the 2D array and the second for-loop on the 1st-index (in respect to the first-loops' index, nodeOne + 1).

That way during the iteration, the second for-loop can do a comparison with the selected index from the first for-loop. In return, duplicates are avoided.